File: llm_written/claude-sonnet-4-5/bfs/go/bfs_parallel.go
// Package main provides a parallel implementation of Breadth-First Search (BFS)
// with deterministic output guarantees.
package main

import (
	"runtime"
	"sort"
	"sync"
)

const (
	// MinVerticesForParallel is the threshold below which sequential BFS is used
	MinVerticesForParallel = 100
	// MinFrontierForParallel is the minimum frontier size for parallel processing
	MinFrontierForParallel = 20
)

// BfsParallel performs level-synchronous parallel BFS
// Returns vertices in BFS order, matching the sequential baseline
func BfsParallel(g Graph, startVertex int) []int {
	if _, exists := g.Vertices[startVertex]; !exists {
		return []int{}
	}

	// Normalize for deterministic iteration
	g.Normalize()

	// Small graph fallback to sequential
	if len(g.Vertices) < MinVerticesForParallel {
		return BfsSequential(g, startVertex)
	}

	visited := &sync.Map{}
	visited.Store(startVertex, true)
	result := []int{}
	currentLevel := []int{startVertex}

	numWorkers := runtime.NumCPU()

	for len(currentLevel) > 0 {
		// Add current level to result
		result = append(result, currentLevel...)

		// Process current level to find next level
		var nextLevel []int
		if len(currentLevel) < MinFrontierForParallel {
			// Sequential processing for small frontiers
			nextLevel = processLevelSequential(g, currentLevel, visited)
		} else {
			// Parallel processing for large frontiers
			nextLevel = processLevelParallel(g, currentLevel, visited, numWorkers)
		}

		// Preserve discovery order (no sorting needed - merge order is deterministic)
		currentLevel = nextLevel
	}

	return result
}

// processLevelSequential processes a BFS level sequentially
func processLevelSequential(g Graph, level []int, visited *sync.Map) []int {
	nextLevel := []int{}
	localVisited := make(map[int]bool)

	for _, vertex := range level {
		neighbors := g.Vertices[vertex]
		for _, neighbor := range neighbors {
			if _, alreadyVisited := visited.Load(neighbor); alreadyVisited {
				continue
			}
			if localVisited[neighbor] {
				continue
			}
			localVisited[neighbor] = true
			visited.Store(neighbor, true)
			nextLevel = append(nextLevel, neighbor)
		}
	}

	return nextLevel
}

// processLevelParallel processes a BFS level in parallel using a worker pool
func processLevelParallel(g Graph, level []int, visited *sync.Map, numWorkers int) []int {
	// Divide work into chunks
	chunkSize := (len(level) + numWorkers - 1) / numWorkers
	if chunkSize < 1 {
		chunkSize = 1
	}

	type chunkResult struct {
		index     int
		neighbors []int
	}

	resultChan := make(chan chunkResult, numWorkers)
	var wg sync.WaitGroup

	// Launch workers to process chunks in parallel
	for i := 0; i < numWorkers; i++ {
		start := i * chunkSize
		if start >= len(level) {
			break
		}
		end := start + chunkSize
		if end > len(level) {
			end = len(level)
		}

		wg.Add(1)
		go func(workerID int, chunk []int) {
			defer wg.Done()

			// Process chunk and collect neighbors in discovery order
			localNext := []int{}
			localVisited := make(map[int]bool)

			for _, vertex := range chunk {
				neighbors := g.Vertices[vertex]
				for _, neighbor := range neighbors {
					if _, alreadyVisited := visited.Load(neighbor); alreadyVisited {
						continue
					}
					if localVisited[neighbor] {
						continue
					}
					localVisited[neighbor] = true
					localNext = append(localNext, neighbor)
				}
			}

			resultChan <- chunkResult{index: workerID, neighbors: localNext}
		}(i, level[start:end])
	}

	// Wait for all workers
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	results := make([]chunkResult, 0, numWorkers)
	for res := range resultChan {
		results = append(results, res)
	}

	// Sort by worker index for deterministic merge order
	sort.Slice(results, func(i, j int) bool {
		return results[i].index < results[j].index
	})

	// Merge results with global deduplication, preserving discovery order
	globalVisited := make(map[int]bool)
	nextLevel := []int{}

	for _, res := range results {
		for _, neighbor := range res.neighbors {
			if globalVisited[neighbor] {
				continue
			}
			if _, alreadyVisited := visited.Load(neighbor); alreadyVisited {
				continue
			}
			globalVisited[neighbor] = true
			visited.Store(neighbor, true)
			nextLevel = append(nextLevel, neighbor)
		}
	}

	return nextLevel
}

File: llm_written/claude-sonnet-4-5/bfs/cpp/bfs_parallel.cpp
#include "bfs_parallel.hpp"
#include <omp.h>
#include <vector>
#include <unordered_set>
#include <algorithm>

std::vector<int> bfs_parallel(Graph& g, int start_vertex) {
    std::vector<int> result;
    
    // Edge case: start vertex not in graph
    if (g.vertices.count(start_vertex) == 0) {
        return result;
    }
    
    // Small graph fallback: use sequential for graphs < 100 vertices
    if (g.vertices.size() < 100) {
        // Sequential BFS
        std::unordered_set<int> visited;
        std::vector<int> current_frontier = {start_vertex};
        visited.insert(start_vertex);
        
        while (!current_frontier.empty()) {
            std::vector<int> next_frontier;
            
            for (int v : current_frontier) {
                result.push_back(v);
                
                for (int neighbor : g.vertices.at(v)) {
                    if (visited.find(neighbor) == visited.end()) {
                        visited.insert(neighbor);
                        next_frontier.push_back(neighbor);
                    }
                }
            }
            
            current_frontier = std::move(next_frontier);
        }
        
        return result;
    }
    
    // Parallel BFS with level synchronization
    // Key: maintain exact sequential order by processing frontier in order
    // but parallelizing the neighbor collection for each vertex
    
    std::unordered_set<int> visited;
    std::vector<int> current_frontier = {start_vertex};
    visited.insert(start_vertex);
    
    int num_threads = omp_get_max_threads();
    
    while (!current_frontier.empty()) {
        // Add current frontier to result (in order)
        result.insert(result.end(), current_frontier.begin(), current_frontier.end());
        
        // For each vertex in frontier, collect its neighbors
        // Store them in per-vertex lists to maintain order
        std::vector<std::vector<int>> per_vertex_neighbors(current_frontier.size());
        
        // Process each vertex in parallel to find its neighbors
        // Use dynamic scheduling with chunk size 1 for better load balancing
        #pragma omp parallel for num_threads(num_threads) schedule(dynamic, 1)
        for (size_t i = 0; i < current_frontier.size(); i++) {
            int current = current_frontier[i];
            const std::vector<int>& neighbors = g.vertices.at(current);
            
            // Collect all neighbors (we'll deduplicate later)
            per_vertex_neighbors[i] = neighbors;
        }
        
        // Sequentially merge and deduplicate to maintain exact order
        std::vector<int> next_frontier;
        
        for (size_t i = 0; i < current_frontier.size(); i++) {
            for (int neighbor : per_vertex_neighbors[i]) {
                if (visited.find(neighbor) == visited.end()) {
                    visited.insert(neighbor);
                    next_frontier.push_back(neighbor);
                }
            }
        }
        
        current_frontier = std::move(next_frontier);
    }
    
    return result;
}
